components/canvas/canvas_paint_thread.rs

```diff
@@ -2,1137 +2,280 @@
-
-            let mut updates = webrender_api::ResourceUpdates::new();
-
-            match self.image_key {
-                Some(image_key) => {
-                    debug!("Updating image {:?}.", image_key);
-                    updates.update_image(image_key,
-                                         descriptor,
-                                         data,
-                                         None);
-                }
-                None => {
-                    self.image_key = Some(self.webrender_api.generate_image_key());
-                    debug!("New image {:?}.", self.image_key);
-                    updates.add_image(self.image_key.unwrap(),
-                                      descriptor,
-                                      data,
-                                      None);
-                }
-            }
-
-            if let Some(image_key) = mem::replace(&mut self.very_old_image_key, self.old_image_key.take()) {
-                updates.delete_image(image_key);
+    fn process_canvas_2d_message(&mut self, message: Canvas2dMsg, canvas_id: CanvasId) {
+        match message {
+            Canvas2dMsg::FillText(text, x, y, max_width) => {
+                self.canvas(&canvas_id).fill_text(text, x, y, max_width)
+            },
+            Canvas2dMsg::FillRect(ref rect) => {
+                self.canvas(&canvas_id).fill_rect(rect)
+            },
+            Canvas2dMsg::StrokeRect(ref rect) => {
+                self.canvas(&canvas_id).stroke_rect(rect)
+            },
+            Canvas2dMsg::ClearRect(ref rect) => {
+                self.canvas(&canvas_id).clear_rect(rect)
+            },
+            Canvas2dMsg::BeginPath => {
+                self.canvas(&canvas_id).begin_path()
+            },
+            Canvas2dMsg::ClosePath => {
+                self.canvas(&canvas_id).close_path()
+            },
+            Canvas2dMsg::Fill => {
+                self.canvas(&canvas_id).fill()
+            },
+            Canvas2dMsg::Stroke => {
+                self.canvas(&canvas_id).stroke()
+            },
+            Canvas2dMsg::Clip => {
+                self.canvas(&canvas_id).clip()
+            },
+            Canvas2dMsg::IsPointInPath(x, y, fill_rule, chan) => {
+                self.canvas(&canvas_id).is_point_in_path(x, y, fill_rule, chan)
+            },
+            Canvas2dMsg::DrawImage(
+                imagedata,
+                image_size,
+                dest_rect,
+                source_rect,
+                smoothing_enabled,
+            ) => {
+                self.canvas(&canvas_id).draw_image(
+                    imagedata.into(),
+                    image_size,
+                    dest_rect,
+                    source_rect,
+                    smoothing_enabled,
+                )
+            },
+            Canvas2dMsg::DrawImageSelf(
+                    image_size,
+                    dest_rect,
+                    source_rect,
+                    smoothing_enabled
```

Comment:
nit: indentation.

Reply:
I totally know how I missed that one. Mixed it up with the next block. Fixed.